osu.Game/Graphics/ParticleExplosion.cs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.OpenGL.Vertices;
+using osu.Framework.Graphics.Primitives;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Graphics.Textures;
+using osu.Framework.Utils;
+using osuTK;
+
+namespace osu.Game.Graphics
+{
+    /// <summary>
+    /// An explosion of textured particles based on how osu-stable randomises the explosion pattern.
+    /// </summary>
+    public class ParticleExplosion : Sprite
+    {
+        private readonly int particleCount;
+        private readonly double duration;
+        private double startTime;
+
+        private readonly List<ParticlePart> parts = new List<ParticlePart>();
+
+        public ParticleExplosion(Texture texture, int particleCount, double duration)
+        {
+            Texture = texture;
+            this.particleCount = particleCount;
+            this.duration = duration;
+            Blending = BlendingParameters.Additive;
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            Restart();
+        }
+
+        /// <summary>
+        /// Restart the animation from the current point in time.
+        /// Supports transform time offset chaining.
+        /// </summary>
+        public void Restart()
+        {
+            startTime = TransformStartTime;
+            this.FadeOutFromOne(duration);
+
+            parts.Clear();
+            for (int i = 0; i < particleCount; i++)
+                parts.Add(new ParticlePart(duration));
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+            Invalidate(Invalidation.DrawNode);
```

Comment:
maybe could add a guard here to not invalidate if `Time.Current <= startTime + duration`? not sure how much of a perf influence that will have, though.

Reply:
I think I'm going to leave this as is just to be cautious. I'm thinking of cases where time jumps in large portions and causes an invalidation to be skipped.

Lifetimes of the explosion should be enough to avoid any performance issues.